yrs/src/types/xml.rs

```diff
@@ -109,421 +132,533 @@ impl XmlElement {
+pub trait Xml: AsRef<Branch> {
+    fn parent(&self) -> Option<XmlElementRef> {
+        let block = self.as_ref().item?;
+        let item = block.as_item()?;
+        let parent = item.parent.as_branch()?;
+        Some(XmlElementRef::from(*parent))
     }
 
-    pub fn to_string<T: ReadTxn>(&self, txn: &T) -> String {
-        let mut s = String::new();
-        let inner = self.inner();
-        for i in inner.iter(txn) {
-            for content in i.content.get_content() {
-                write!(&mut s, "{}", content.to_string(txn)).unwrap();
+    /// Removes an attribute recognized by an `attr_name` from a current XML element.
+    fn remove_attribute<K>(&self, txn: &mut TransactionMut, attr_name: &K)
+    where
+        K: AsRef<str>,
+    {
+        self.as_ref().remove(txn, attr_name.as_ref());
+    }
+
+    /// Inserts an attribute entry into current XML element.
+    fn insert_attribute<K, V>(&self, txn: &mut TransactionMut, attr_name: K, attr_value: V)
+    where
+        K: Into<Rc<str>>,
+        V: AsRef<str>,
+    {
+        let key = attr_name.into();
+        let value = crate::block::PrelimString(attr_value.as_ref().into());
+        let pos = {
+            let branch = self.as_ref();
+            let left = branch.map.get(&key);
+            ItemPosition {
+                parent: BranchPtr::from(branch).into(),
+                left: left.cloned(),
+                right: None,
+                index: 0,
+                current_attrs: None,
             }
-        }
-        s
+        };
+
+        txn.create_item(&pos, value, Some(key));
+    }
+
+    /// Returns a value of an attribute given its `attr_name`. Returns `None` if no such attribute
+    /// can be found inside of a current XML element.
+    fn get_attribute<T: ReadTxn>(&self, txn: &T, attr_name: &str) -> Option<String> {
+        let branch = self.as_ref();
+        let value = branch.get(txn, attr_name)?;
+        Some(value.to_string(txn))
+    }
+
+    /// Returns an unordered iterator over all attributes (key-value pairs), that can be found
+    /// inside of a current XML element.
+    fn attributes<'a, T: ReadTxn>(&'a self, txn: &'a T) -> Attributes<'a, &'a T, T> {
+        Attributes(Entries::new(&self.as_ref().map, txn))
     }
 
-    pub fn insert_elem<S: Into<Rc<str>>>(
-        &self,
-        txn: &mut TransactionMut,
-        index: u32,
-        name: S,
-    ) -> XmlElement {
-        let ptr = self.0.insert_at(txn, index, PrelimXml::Elem(name.into()));
-        let item = ptr.as_item().unwrap();
-        if let ItemContent::Type(inner) = &item.content {
-            XmlElement::from(BranchPtr::from(inner))
-        } else {
-            panic!("Defect: inserted XML element returned primitive value block")
+    fn siblings<'a, T: ReadTxn>(&self, txn: &'a T) -> Siblings<'a, T> {
+        let ptr = BranchPtr::from(self.as_ref());
+        Siblings::new(ptr.item, txn)
+    }
+}
+
+pub trait XmlFragment: AsRef<Branch> {
```

Comment:
`XmlFragment` trait API is similar to what `Array` API has to offer, but it's focused solely on Xml nodes (things that implement `XmlPrelim` trait). 

Now we no longer need to have separate methods like: `xml.push_elem_back`/`xml.push_text_back`. Instead we can use prelim types, eg. `xml.push_back(txn, XmlElementPrelim([]))` or `xml.push_back(txn, XmlTextPrelim(""))`.